spec/models/file_upload_handler_spec.rb

```diff
@@ -0,0 +1,79 @@
+# frozen_string_literal: true
+
+require 'spec_helper'
+
+describe 'FileUploadHandler' do
+  let(:scenario) { FactoryBot.create(:scenario) }
+  let(:file_handler) { FileUploadHandler.new(file, 'my_esdl', key, scenario) }
+
+  context 'when a file was already attached' do
+    let(:key) { 'esdl_file' }
+    let(:file) { fixture_file_upload('files/valid_esdl_file.esdl', 'text/xml').read }
+
+    before { file_handler.call }
+
+    it 'is not valid' do
+      expect(file_handler).not_to be_valid
+    end
+  end
+
+  context 'when attaching a valid esdl file' do
+    let(:key) { 'esdl_file' }
+    let(:file) { fixture_file_upload('files/valid_esdl_file.esdl', 'text/xml').read }
+
+    it 'is valid' do
+      expect(file_handler).to be_valid
+    end
+
+    it 'attaches the file' do
+      expect { file_handler.call }
+        .to change {
+          scenario
+            .reload
+            .attachments
+            .find_by(key: 'esdl_file')
+            .present?
+        }
+        .from(false).to(true)
+    end
+  end
+
+  # We don't really care - should we?
```

Comment:
@antw do you think we should ensure the file is ESDL/xml before we allow it to be attached? Only the etm-esdl app uses this feature right now, so that kind of ensures the file being a valid ESDL file. However, the API is open to everyone that wants to use it.

Reply:
Sure, we could do.

One naive approach is just to read the first 1kb or so, and check that there's something that looks like an `edsl:EnergySystem` tag:

```
header = File.open(path_to_uploaded_tempfile) { |file| file.gets(nil, 1024) }

unless header.include?('<esdl:EnergySystem')
  errors.add('...')
end
```

It's not exactly foolproof (the file could still not be valid ESDL), but it might help catch some user errors.

If you wanted, we could do more detailed validations by loading the XML file with Nokogiri and checking for the existence of certain tags or attributes, but I'm not sure if you think that's worth it?

I feel as though checking for the esdl:EnergySystem tag in the first 1k is probably good enough.

Reply:
> I feel as though checking for the esdl:EnergySystem tag in the first 1k is probably good enough.

Agreed! The fact that it is an ESDL file is more important than it being valid ESDL, in my opinion. We would have to keep up with new ESDL versions coming out every few months, doing that in just the etm-esdl app is hard enough. 

Thanks for your code snippet!